### ISSUE TITLE
Http Client performance drop in preview-2

### ISSUE THREAD

**Issue by Petermarcu** [dotnet/corefx#29065]
@rahul10992 commented on [Thu Apr 12 2018](https://github.com/dotnet/core/issues/1437)

Hello. I was trying out HttpClient benchmarking for net core 2.1 and well, I am seeing a serious drop in the HttpClient provided in preview 2.

for simple Get requests between 2 Windows Server 2016 VMs set up with network acceleration in Azure, 
1. for Preview-1 with the "AppContext.SetSwitch("System.Net.Http.UseManagedHttpClientHandler", true);" I get 75,000 requests/second
2. For preview-2 using the SocketsHttpHandler, the performance drops to 43,000 requests/second

Is there a reason for this performance drop?

**Comment by Petermarcu** [comment]
@geoffkizer @karelz

**Comment by karelz** [comment]
@rahul10992 can you please share your perf repro?
We are not aware of any intentional perf drops between Preview1 and Preview2.

**Comment by rahul10992** [comment]
Hello. What I used to test this out is here: https://github.com/rahul10992/net-core-preview-issue 
You will have to enter the right IP Address in the HttpClientTester's Program.cs to ping the right endpoint. But apart from that, this should work. 

As for my azure environment,
I created 2 DS4_V2 Windows Server 2016 VMs with network acceleration on the same virtual network and had to open up the 5000 port on the side that hosts the endpoint. I also installed preview 1 first for preview 1 testing and then preview 2.

**Comment by karelz** [comment]
@caesar1995 can you please try the repro locally on our perf machines? Let's see if we can reproduce it.

**Comment by caesar1995** [comment]
I have tested the repro with loopback address. The repro is testing `SocketsHttpHandler` sending GET request to a Kestrel server.

The variance is on the server side: `Microsoft.AspNetCore` targeting 2.1.0-preview1-final/2.1.0-preview2-final. To be clear, it's NOT about testing the `SocketsHttpHandler` targeting preview1/preview2. (So if there is a regression, it should not be on the client side).

The repro works fine when  `Microsoft.AspNetCore` version is 2.1.0-preview2. The rate is steady around 3300 requests/second.

```
C:\HttpBenchmarking\Preview 2\HttpClientTester>dotnet run
Duration: 55 seconds, Total: 184943, Rate: 3362
```

However, if target `Microsoft.AspNetCore` to 2.1.0-preview1, I will have following error on server side:

```
C:\HttpBenchmarking\Preview 1\HttpEndpoint>dotnet run

Unhandled Exception: System.TypeLoadException: Method 'Rent' in type 'System.Buffers.MemoryPool' from assembly 'System.IO.Pipelines, Version=0.1.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' does not have an implementation.
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransport..ctor(IEndPointInformation endPointInformation, IConnectionHandler handler, IApplicationLifetime applicationLifetime, ISocketsTrace trace)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransportFactory.Create(IEndPointInformation endPointInformation, IConnectionHandler handler)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer.<>c__DisplayClass22_0`1.<<StartAsync>g__OnBind|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindEndpointAsync(ListenOptions endpoint, AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions.BindAsync(AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.AnyIPListenOptions.BindAsync(AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.AddressesStrategy.BindAsync(AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindAsync(IServerAddressesFeature addresses, KestrelServerOptions serverOptions, ILogger logger, Func`2 createBinding)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer.StartAsync[TContext](IHttpApplication`1 application, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.StartAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Hosting.WebHostExtensions.RunAsync(IWebHost host, CancellationToken token, String shutdownMessage)
   at Microsoft.AspNetCore.Hosting.WebHostExtensions.RunAsync(IWebHost host, CancellationToken token)
   at Microsoft.AspNetCore.Hosting.WebHostExtensions.Run(IWebHost host)
   at HttpEndpoint.Program.Main(String[] args) in C:\Users\shizchen\Desktop\net-core-preview-issue-master\net-core-preview-issue-master\HttpBenchmarking\Preview 1\HttpEndpoint\Program.cs:line 25
```

`dotnet --info` gives me:

```
.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0-preview1-final [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.0-preview2-final [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0-preview1-final [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.0-preview2-final [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0-preview1-26216-03 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0-preview2-26406-04 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
```

I tried following commands `dotnet add package`:

```
C:\Users\shizchen\Desktop\net-core-preview-issue-master\net-core-preview-issue-master\HttpBenchmarking\Preview 1\HttpEndpoint>dotnet add package Microsoft.AspNetCore.Server.Kestrel --version 2.1.0-preview1-final
  Writing C:\Users\shizchen\AppData\Local\Temp\tmp8960.tmp
info : Adding PackageReference for package 'Microsoft.AspNetCore.Server.Kestrel' into project 'C:\Users\shizchen\Desktop\net-core-preview-issue-master\net-core-preview-issue-master\HttpBenchmarking\Preview 1\HttpEndpoint\HttpEndpoint.csproj'.
log  : Restoring packages for C:\Users\shizchen\Desktop\net-core-preview-issue-master\net-core-preview-issue-master\HttpBenchmarking\Preview 1\HttpEndpoint\HttpEndpoint.csproj...
info : Package 'Microsoft.AspNetCore.Server.Kestrel' is compatible with all the specified frameworks in project 'C:\Users\shizchen\Desktop\net-core-preview-issue-master\net-core-preview-issue-master\HttpBenchmarking\Preview 1\HttpEndpoint\HttpEndpoint.csproj'.
info : PackageReference for package 'Microsoft.AspNetCore.Server.Kestrel' version '2.1.0-preview1-final' updated in file 'C:\Users\shizchen\Desktop\net-core-preview-issue-master\net-core-preview-issue-master\HttpBenchmarking\Preview 1\HttpEndpoint\HttpEndpoint.csproj'.

C:\Users\shizchen\Desktop\net-core-preview-issue-master\net-core-preview-issue-master\HttpBenchmarking\Preview 1\HttpEndpoint>dotnet add package Microsoft.AspNetCore.Hosting --version 2.1.0-preview1-final
  Writing C:\Users\shizchen\AppData\Local\Temp\tmp57EB.tmp
info : Adding PackageReference for package 'Microsoft.AspNetCore.Hosting' into project 'C:\Users\shizchen\Desktop\net-core-preview-issue-master\net-core-preview-issue-master\HttpBenchmarking\Preview 1\HttpEndpoint\HttpEndpoint.csproj'.
log  : Restoring packages for C:\Users\shizchen\Desktop\net-core-preview-issue-master\net-core-preview-issue-master\HttpBenchmarking\Preview 1\HttpEndpoint\HttpEndpoint.csproj...
info : Package 'Microsoft.AspNetCore.Hosting' is compatible with all the specified frameworks in project 'C:\Users\shizchen\Desktop\net-core-preview-issue-master\net-core-preview-issue-master\HttpBenchmarking\Preview 1\HttpEndpoint\HttpEndpoint.csproj'.
info : PackageReference for package 'Microsoft.AspNetCore.Hosting' version '2.1.0-preview1-final' updated in file 'C:\Users\shizchen\Desktop\net-core-preview-issue-master\net-core-preview-issue-master\HttpBenchmarking\Preview 1\HttpEndpoint\HttpEndpoint.csproj'.
```

Then do `dotnet restore` `dotnet run`, still fails with `System.TypeLoadException`.

I will need some guidance from AspNetCore side to continue reproing the issue.

/cc: @karelz @davidsh

**Comment by stephentoub** [comment]
@rahul10992, just a guess, could you try setting UseProxy to false?

**Comment by caesar1995** [comment]
Hey Stephen, UseProxy definitely contributes to the regression. (Hopefully #29050 resolves it).

FWIW, here is the data from preview 2 with the repro using our perf machine: 
<img width="759" alt="capture" src="https://user-images.githubusercontent.com/8537784/38709861-3161a334-3e72-11e8-9498-3c09d0781994.PNG">

The first three run didn't specify `UseProxy` to false. The average is: (45190+47349+47786)/3 = 46775 requests/second.
The last three run with `UseProxy = false`. The average is: (56341+55982+58941)/3 = 57088 requests/second.

This accounts for (57088-46775)/57088 = 18% performance regression.

I didn't verify against preview 1 (have some weird setup issue), but from issue description, we have (75000-43,000)/75000 = 42% performance regression. I will continue the preview 1 setup, so that we'll have baseline data.

**Comment by caesar1995** [comment]
Updated client & server side with Release flag. No differnece. So I guess `dotnet run` run with Release compilation.

Notice that in order to make sure it's client side issue, on the server side I always target the with preview 2, and didn't switch back and forth. (Client: preview 1/preview 2, Server: preview 2).

From the description/repro, it looks like @rahul10992 is testing: 
Client: preview 1 -> Server: preview 1
Client: preview 2 -> Server: preview 2.

I will verify Client: preview 1 -> Server: preview 1 case.

**Comment by karelz** [comment]
I don't think testing client & server both on Preview 1 is priority. Let's dig deeper into the ~13% remaining difference (against UseProxy=false).
I think it would be best to collect PerfView traces and analyze the diff.

**Comment by caesar1995** [comment]
Seems on server side, preview 2 didn't cause regression. Below are the test I did against Server: Preview 1.

Client Preview 1: average 65933 requests/seconds.
Client Preview 2: average 49301 requests/seconds.
Client Preview 2 with `UseProxy = false`: average 58176 requests/seconds.

Comparing to Server: Preview 2:

Client Preview 1: average 68923 requests/seconds.
Client Preview 2: average 46775 requests/seconds.
Client Preview 2 with `UseProxy = false`: average 57088 requests/seconds.

No too much difference form different server version. So it's mainly regression in SocketsHttpHandler.

**Comment by stephentoub** [comment]
Just to make sure it's not an issue we've already fixed, could you also try using the latest from master?

**Comment by rahul10992** [comment]
Hello. My message is late. @stephentoub  But I tried it with UseProxy = false and the performance for preview-2 jumps up to 74,000 req/sec.

**Comment by stephentoub** [comment]
> I tried it with UseProxy = false and the performance for preview-2 jumps up to 74,000 req/sec.

Thanks, @rahul10992.  So it sounds like setting that to false basically reversed the regression, right?  The cause for that is we added support for automatic proxy discovery.  But we implemented a workaround in https://github.com/dotnet/corefx/pull/29050 that should basically behave perf-wise as if UseProxy=false if you don't have a proxy auto-discovery file.  And of course you can explicitly set UseProxy to false to entirely avoid any auto-discovery.

**Comment by rahul10992** [comment]
Thanks @stephentoub! I will keep this in mind for my future tests!

**Comment by karelz** [comment]
@caesar1995 can you please patch nightly build with fix of #29050 as we discussed in standup? (just grab latest binary with the fix and replace it in the latest nightly build)
It will help us double-check that fix for #29050 helps in the end-to-end scenario and makes the regression go away. Thanks!

**Comment by caesar1995** [comment]
I tried using my local corefx build for the repro, there are some noise which I'm not sure that if #29050 resolved the regression.

I will wait until Monday to use the next official nightly build to confirm it.

**Comment by caesar1995** [comment]
I can confirm that #29050 resolved the regression with 2.1.300-preview3-008627.

Without setting `UseProxy = false`, 55235 requests/second.
With `UseProxy = false`, 55878 requests/second.

<img width="516" alt="capture" src="https://user-images.githubusercontent.com/8537784/38783897-ee736a28-40bd-11e8-98e1-e37f34b82cb6.PNG">

To entirely avoid any auto-discovery, we can explicitly set UseProxy to false. Closing the issue.

**Comment by karelz** [comment]
Thanks @caesar1995!
